File: src/asqav/extras/openai_agents.py
"""OpenAI Agents SDK integration for asqav.

Install: pip install asqav[openai-agents]

Provides two integration points:

- ``AsqavGuardrail`` - signs input/output events on agent runs.
- ``AsqavTracingProcessor`` - signs trace and span lifecycle events
  as governance records via the OpenAI Agents tracing pipeline.

Usage::

    from asqav.extras.openai_agents import AsqavGuardrail, AsqavTracingProcessor
    from agents import Agent, add_trace_processor

    # Guardrail (audit-only, never blocks)
    agent = Agent(name="my-agent", guardrails=[AsqavGuardrail(agent_name="my-openai-agent")])

    # Tracing processor (signs every trace/span lifecycle event)
    add_trace_processor(AsqavTracingProcessor(agent_name="my-openai-agent"))
"""

from __future__ import annotations

import dataclasses
import logging
from typing import Any

try:
    from agents.tracing import Span, Trace, TracingProcessor
except ImportError:
    raise ImportError(
        "OpenAI Agents integration requires openai-agents. "
        "Install with: pip install asqav[openai-agents]"
    )

from ._base import AsqavAdapter

logger = logging.getLogger("asqav")


@dataclasses.dataclass
class GuardrailResult:
    """Lightweight result matching the OpenAI Agents SDK guardrail interface.

    Defined locally so the integration does not break when the SDK's
    internal types change between versions (adapter pattern).
    """

    passed: bool
    output: Any | None = None


class AsqavGuardrail(AsqavAdapter):
    """Guardrail that signs input/output events on OpenAI agent runs.

    Asqav guardrails are audit-only: they sign governance events but never
    block agent execution. Both ``run_input_guardrail`` and
    ``run_output_guardrail`` always return ``GuardrailResult(passed=True)``.

    Args:
        api_key: Optional API key override (uses ``asqav.init()`` default).
        agent_name: Name for a new asqav agent (calls ``Agent.create``).
        agent_id: ID of an existing asqav agent (calls ``Agent.get``).
    """

    def _resolve_agent_name(self, agent: Any) -> str:
        """Extract agent name from the OpenAI Agent object."""
        name = getattr(agent, "name", None)
        if name:
            return str(name)
        try:
            return str(agent)
        except Exception:
            return "unknown"

    async def run_input_guardrail(
        self, agent: Any, input_data: Any
    ) -> GuardrailResult:
        """Sign an agent:input governance event.

        Called by the OpenAI Agents SDK before an agent processes input.
        Signing failures are swallowed (fail-open) so the agent run
        is never interrupted by governance issues.
        """
        try:
            agent_name = self._resolve_agent_name(agent)
            data_repr = repr(input_data)
            if len(data_repr) > 200:
                data_repr = data_repr[:200]
            self._sign_action(
                "agent:input",
                {
                    "agent_name": agent_name,
                    "input_type": type(input_data).__name__,
                    "input_length": len(data_repr),
                    "input_preview": data_repr,
                },
            )
        except Exception as exc:
            logger.warning("asqav input guardrail signing failed (fail-open): %s", exc)
        return GuardrailResult(passed=True, output=None)

    async def run_output_guardrail(
        self, agent: Any, output_data: Any
    ) -> GuardrailResult:
        """Sign an agent:output governance event.

        Called by the OpenAI Agents SDK after an agent produces output.
        Signing failures are swallowed (fail-open) so the agent run
        is never interrupted by governance issues.
        """
        try:
            agent_name = self._resolve_agent_name(agent)
            data_repr = repr(output_data)
            if len(data_repr) > 200:
                data_repr = data_repr[:200]
            self._sign_action(
                "agent:output",
                {
                    "agent_name": agent_name,
                    "output_type": type(output_data).__name__,
                    "output_length": len(data_repr),
                    "output_preview": data_repr,
                },
            )
        except Exception as exc:
            logger.warning("asqav output guardrail signing failed (fail-open): %s", exc)
        return GuardrailResult(passed=True, output=None)


class AsqavTracingProcessor(AsqavAdapter, TracingProcessor):
    """Tracing processor that signs trace and span lifecycle events.

    Registers with the OpenAI Agents tracing pipeline via
    ``add_trace_processor``. Each trace start/end and span start/end
    is signed as a governance event through asqav.

    All signing is fail-open - tracing failures are logged but never
    raise, so the agent pipeline is never interrupted.

    Args:
        api_key: Optional API key override (uses ``asqav.init()`` default).
        agent_name: Name for a new asqav agent (calls ``Agent.create``).
        agent_id: ID of an existing asqav agent (calls ``Agent.get``).
    """

    def on_trace_start(self, trace: Trace) -> None:
        """Sign a trace:start governance event."""
        try:
            self._sign_action(
                "trace:start",
                {
                    "trace_id": trace.trace_id,
                    "trace_name": trace.name,
                },
            )
        except Exception as exc:
            logger.warning("asqav trace:start signing failed (fail-open): %s", exc)

    def on_trace_end(self, trace: Trace) -> None:
        """Sign a trace:end governance event."""
        try:
            self._sign_action(
                "trace:end",
                {
                    "trace_id": trace.trace_id,
                    "trace_name": trace.name,
                },
            )
        except Exception as exc:
            logger.warning("asqav trace:end signing failed (fail-open): %s", exc)

    def on_span_start(self, span: Span[Any]) -> None:
        """Sign a span:start governance event."""
        try:
            span_data = span.span_data
            self._sign_action(
                "span:start",
                {
                    "trace_id": span.trace_id,
                    "span_id": span.span_id,
                    "span_type": type(span_data).__name__,
                },
            )
        except Exception as exc:
            logger.warning("asqav span:start signing failed (fail-open): %s", exc)

    def on_span_end(self, span: Span[Any]) -> None:
        """Sign a span:end governance event."""
        try:
            span_data = span.span_data
            exported = span.export() if hasattr(span, "export") else None
            context: dict[str, Any] = {
                "trace_id": span.trace_id,
                "span_id": span.span_id,
                "span_type": type(span_data).__name__,
            }
            if exported:
                # Include a preview of the span output for the audit record
                output = exported.get("span_data", {}).get("output")
                if output is not None:
                    output_repr = repr(output)
                    if len(output_repr) > 200:
                        output_repr = output_repr[:200]
                    context["output_preview"] = output_repr
            self._sign_action("span:end", context)
        except Exception as exc:
            logger.warning("asqav span:end signing failed (fail-open): %s", exc)

    def shutdown(self) -> None:
        """End the asqav session on processor shutdown."""
        self._end_session(status="completed")

    def force_flush(self) -> None:
        """No-op - asqav signs events synchronously."""
        pass

File: typescript/src/extras/openai-agents.ts
/**
 * OpenAI Agents JS integration for Asqav.
 *
 * Install (peer dependency):
 *   npm install @openai/agents
 *
 * Usage:
 *   import { init } from "@asqav/sdk";
 *   import { AsqavOpenAIAgentsAdapter } from "@asqav/sdk/extras/openai-agents";
 *   init({ apiKey: process.env.ASQAV_API_KEY! });
 *   const adapter = new AsqavOpenAIAgentsAdapter({ agentName: "my-agent" });
 *   adapter.wrapTool(myTool);
 *
 * Mirrors the Python ``asqav.extras.openai_agents.AsqavGuardrail`` /
 * ``AsqavTracingProcessor`` adapter: signs ``agent:input`` /
 * ``agent:output`` / ``tool:start`` / ``tool:end`` events as the agent
 * runs. Operates fail-open so governance never breaks the agent loop.
 */

import { AsqavAdapter, raiseMissingPeer, type AsqavAdapterOptions } from "./_base.js";

const MAX_PREVIEW = 200;

function truncate(s: string, max = MAX_PREVIEW): string {
  return s.length > max ? s.slice(0, max) : s;
}

/**
 * Minimal surface of an OpenAI Agents JS ``tool``. We accept any object
 * with ``name`` and ``execute`` so we never bind to a specific SDK
 * version.
 */
export interface OpenAIAgentsToolLike<I = unknown, O = unknown> {
  name?: string;
  description?: string;
  execute?: (input: I, ...rest: unknown[]) => Promise<O> | O;
}

export interface AsqavOpenAIAgentsAdapterOptions extends AsqavAdapterOptions {
  /** Optional adapter display name. */
  name?: string;
}

/**
 * Wraps OpenAI Agents JS tools and run-loop callbacks so every
 * tool-call lifecycle event flows through Asqav. Mirrors the Python
 * ``AsqavGuardrail`` and ``AsqavTracingProcessor``.
 */
export class AsqavOpenAIAgentsAdapter extends AsqavAdapter {
  public readonly name: string;

  constructor(options: AsqavOpenAIAgentsAdapterOptions) {
    super(options);
    this.name = options.name ?? "asqav-openai-agents-adapter";
  }

  /** Verify ``@openai/agents`` is installed; throw a clear error if not. */
  static async loadPeer(): Promise<unknown> {
    try {
      // Optional peer; not declared in normal dependencies.
      // @ts-ignore optional peer dependency
      return (await import("@openai/agents")) as unknown;
    } catch (err) {
      raiseMissingPeer(
        "OpenAI Agents JS",
        "@openai/agents",
        "npm install @openai/agents",
        err,
      );
    }
  }

  /**
   * Wrap an OpenAI Agents tool so each invocation signs
   * ``tool:start`` and ``tool:end`` (or ``tool:error``). Returns a new
   * tool object with the same ``name``/``description`` and a wrapped
   * ``execute``.
   */
  wrapTool<I, O>(tool: OpenAIAgentsToolLike<I, O>): OpenAIAgentsToolLike<I, O> {
    if (typeof tool.execute !== "function") {
      throw new Error("wrapTool: tool.execute must be a function");
    }
    const original = tool.execute.bind(tool);
    const self = this;
    const toolName = tool.name ?? "unknown";

    const execute = async (input: I, ...rest: unknown[]): Promise<O> => {
      const inputStr = typeof input === "string" ? input : JSON.stringify(input);
      self.signAction({
        actionType: "tool:start",
        context: {
          tool: String(toolName),
          input_preview: truncate(inputStr ?? ""),
        },
      });
      try {
        const out = await original(input, ...rest);
        const outStr = typeof out === "string" ? out : JSON.stringify(out);
        self.signAction({
          actionType: "tool:end",
          context: {
            tool: String(toolName),
            output_length: outStr?.length ?? 0,
            output_preview: truncate(outStr ?? ""),
          },
        });
        return out;
      } catch (err) {
        self.signAction({
          actionType: "tool:error",
          context: {
            tool: String(toolName),
            error_type: err instanceof Error ? err.name : typeof err,
            error: truncate(err instanceof Error ? err.message : String(err)),
          },
        });
        throw err;
      }
    };

    return { ...tool, execute };
  }

  /** Sign an ``agent:input`` event. Use as an input guardrail. */
  public onAgentInput(input: unknown, agentName?: string): void {
    const repr = typeof input === "string" ? input : JSON.stringify(input);
    this.signAction({
      actionType: "agent:input",
      context: {
        agent_name: agentName ?? this.name,
        input_type: typeof input,
        input_length: repr?.length ?? 0,
        input_preview: truncate(repr ?? ""),
      },
    });
  }

  /** Sign an ``agent:output`` event. Use as an output guardrail. */
  public onAgentOutput(output: unknown, agentName?: string): void {
    const repr = typeof output === "string" ? output : JSON.stringify(output);
    this.signAction({
      actionType: "agent:output",
      context: {
        agent_name: agentName ?? this.name,
        output_type: typeof output,
        output_length: repr?.length ?? 0,
        output_preview: truncate(repr ?? ""),
      },
    });
  }
}
